Add clear-cache

rikhuijzer opened the pull request:
This PR is an attempt to implement clearing of the cache (https://github.com/JuliaCollections/Memoize.jl/issues/14). 

EDIT: In a previous version of this comment, I stated that other types than `IdDict` needed to be considered. Since `Base.empty!` takes collections, I believe that the code should now work for any relevant type.

cstjean commented on src/Memoize.jl:
```diff
@@ -76,4 +76,17 @@ macro memoize(args...)
+    mod = __module__
+    fcache = getfield(mod, fcachename)
+
+    function clear_cache!(x::IdDict)
+        foreach(k -> delete!(fcache, k), keys(fcache))
+    end
+
+    Core.eval(mod, :($(clear_cache!(fcache))))
```
It's generally preferable to have macros that expand into code that does the work, rather than do the work in the macro itself. Perhaps this would work:

```suggestion
    esc(:(Base.empty!(fcachename)))
```
?

rikhuijzer replied:
Great suggestion to use `Base.empty!`!

cstjean commented on src/Memoize.jl:
```diff
@@ -78,15 +78,10 @@ macro memoize(args...)
-    end
-
-    Core.eval(mod, :($(clear_cache!(fcache))))
+    esc(quote
+        fcachename = Symbol("##", $f, "_memoized_cache")
+        Base.empty!(eval(fcachename))
+    end)
```
```suggestion
    fcachename = Symbol("##", $f, "_memoized_cache")
    esc(quote
        Base.empty!($fcachename)
    end)
```
The less `eval` the better...

cstjean commented on src/Memoize.jl:
```diff
@@ -76,4 +76,12 @@ macro memoize(args...)
+macro clear_cache(f)
+    fcachename = Symbol("##", f, "_memoized_cache")
+    esc(quote
+        Base.empty!(eval($fcachename))
```
```suggestion
        Base.empty!($fcachename)
```

rikhuijzer replied:
Apologies for misreading the suggestion!

rikhuijzer commented:
It seems that this PR is not making progress. Is there anything I can do?

cstjean commented:
If you do

```julia
module A
@memoize foo(x) = x
end

using A: foo

@clear_memoize_cache foo
```

I don't believe it would work, right? I'm sorry that I didn't think of it earlier.

I think that `clear_memoize_cache!(foo)` would be the better approach. That means you need to implement `Memoize.clear_memoize_cache(::typeof(foo))` in the `@memoize` expansion.

rikhuijzer commented:
Thank you for looking into it again. Well spotted; no worries, I should have noticed that the functionality wasn't handling modules properly.

I've changed it to `empty!(@memoize_cache(foo))` as you've asked. It is a macro call because the module of the caller has to be determined. 

The new functionality is also documented again in the README. I've changed `a` to `2a` to change 
```
 julia> @memoize_cache(x)
  IdDict{Any,Any} with 1 entry:
    (1,) => 1
```
into 
```
 julia> @memoize_cache(x)
  IdDict{Any,Any} with 1 entry:
    (1,) => 2
```
which I think is more clear to readers.

cstjean commented:
> I've changed it to empty!(@memoize_cache(foo)) as you've asked. It is a macro call because the module of the caller has to be determined.

I don't think that would work in another module. Try the example I gave:

```julia
module A
@memoize foo(x) = x
end

using A: foo
@clear_memoize_cache foo
```

If you implement `memoize_cache(::typeof(foo))`, we'll get the behaviour we want, which is that the cache is associated to the _function_ `foo`, and not to the `Symbol` `foo`.

rikhuijzer commented:
Isn't that what is tested with
```
  using .MemoizeTest
  using .MemoizeTest: custom_dict

  empty!(@memoize_cache(custom_dict))
  @test custom_dict(1) == 1
  @test MemoizeTest.run == 3
  @test custom_dict(1) == 1
  @test MemoizeTest.run == 3

  empty!(@memoize_cache(MemoizeTest.custom_dict))
  @test custom_dict(1) == 1
  @test MemoizeTest.run == 4
```
in the tests? If I understand correctly, this is the example you gave
```
julia> using Memoize

julia> module A
       using Memoize
       @memoize foo(x) = x
       end
Main.A

julia> using .A: foo

julia> @memoize_cache(foo)
IdDict{Any,Any}()

julia> foo(3)
3

julia> @memoize_cache(foo)
IdDict{Any,Any} with 1 entry:
  (3,) => 3

julia> empty!(@memoize_cache(foo))
IdDict{Any,Any}()
```

rikhuijzer commented:
Oh, and it seems that this is failing on Julia 1.1 with
```
ERROR: LoadError: "Main.MemoizeTest.custom_dict" is not defined in module Main
```
It seems that I need to fix that.

rikhuijzer commented:
On second thought, I'm not going to invest time on Julia 1.1. I did that at https://github.com/GiovineItalia/Gadfly.jl/pull/1496 and it was not a good time investment.

Let me know if I can make changes to improve the functionality on Julia 1.5. If Julia 1.1 is a must, then feel free to close this PR.

cstjean commented on src/Memoize.jl:
```diff
@@ -76,4 +78,18 @@ macro memoize(args...)
+
+function memoize_cache(m::Module, f::Function)
+    fmodulename = which(m, Symbol(f))
+    getproperty(fmodulename, cache_name(f))
```
```suggestion
    getproperty(parentmodule(f), cache_name(f))
```
Would that work? Then you can get rid of the `m` argument entirely, and `memoize_cache(f)` becomes a plain function.

It'll fail in certain circumstances (eg. `@memoize Base.sin(::MyNumberType) = ...`) but I don't think there's a clean answer here, because you can already have multiple caches for `Base.sin` if the methods are defined in different modules.

rikhuijzer replied:
Awesome suggestion! :smile:

cstjean reviewed (commented):
I wouldn't mind dropping 1.1 if it makes things easier and it's necessary.

cstjean commented:
> in the tests? If I understand correctly, this is the example you gave

Right, sorry that I didn't read the code properly.

rikhuijzer commented:
> Right, sorry that I didn't read the code properly.

No problem; it happens

I've now used `parentmodule`, which is great! :rocket:  In Julia 1.1, I get
```
ERROR: LoadError: UndefVarError: ##Main.MemoizeTest.custom_dict_memoized_cache not defined
```
In Julia 1.2 the tests pass, so I've bumped the Julia version to 1.2.

cstjean commented on src/Memoize.jl:
```diff
@@ -76,4 +78,9 @@ macro memoize(args...)
 
 end
+
+function memoize_cache(f::Function)
```
```suggestion
function memoize_cache(f::Function)
    # This will fail in certain circumstances (eg. @memoize Base.sin(::MyNumberType) = ...) but I don't think there's 
    # a clean answer here, because we can already have multiple caches for certain functions, if the methods are 
    # defined in different modules.
```

cstjean commented:
Will merge when the checks are green again.
